stomp_plugins/include/stomp_plugins/noise_generators/cartesian_constraints_sampling.h

```diff
@@ -0,0 +1,147 @@
+   * @param parameters        The current value of the optimized parameters [num_dimensions x num_parameters]
+   * @param start_timestep    Start index into the 'parameters' array, usually 0.
+   * @param num_timesteps     The number of elements to use from 'parameters' starting from 'start_timestep'
+   * @param iteration_number  The current iteration count in the optimization loop
+   * @param rollout_number    The index of the noisy trajectory.
+   * @param parameters_noise  The parameters + noise
+   * @param noise             The noise applied to the parameters
+   * @return true if cost were properly computed, false otherwise.
+   */
+  virtual bool generateNoise(const Eigen::MatrixXd& parameters,
+                                       std::size_t start_timestep,
+                                       std::size_t num_timesteps,
+                                       int iteration_number,
+                                       int rollout_number,
+                                       Eigen::MatrixXd& parameters_noise,
+                                       Eigen::MatrixXd& noise) override;
+
+  /**
+   * @brief Called by the Stomp at the end of the optimization process
+   *
+   * @param success           Whether the optimization succeeded
+   * @param total_iterations  Number of iterations used
+   * @param final_cost        The cost value after optimizing.
+   * @param parameters        The parameters generated at the end of current iteration[num_dimensions x num_timesteps]
+   */
+  virtual void done(bool success,int total_iterations,double final_cost,const Eigen::MatrixXd& parameters){}
+
+
+  virtual std::string getName() const
+  {
+    return name_ + "/" + group_;
+  }
+
+
+  virtual std::string getGroupName() const
+  {
+    return group_;
+  }
+
+protected:
+
+  virtual bool setupNoiseGeneration(const planning_scene::PlanningSceneConstPtr& planning_scene,
+                   const moveit_msgs::MotionPlanRequest &req,
+                   const stomp_core::StompConfiguration &config,
+                   moveit_msgs::MoveItErrorCodes& error_code);
+
+  virtual bool setupRobotState(const planning_scene::PlanningSceneConstPtr& planning_scene,
+                   const moveit_msgs::MotionPlanRequest &req,
+                   const stomp_core::StompConfiguration &config,
+                   moveit_msgs::MoveItErrorCodes& error_code);
+  /**
+   * @brief Genereates a random tool pose by apply noise on the redundant axis to a reference tool pose;
+   * @param reference_joint_pose  Joint position used in computing the reference tool pose with FK
+   * @param result       The joint position corresponding to the randomized tool pose
+   * @return noisy tool position
+   */
+  virtual Eigen::Affine3d applyCartesianNoise(const Eigen::VectorXd& reference_joint_pose);
+
+protected:
+
+  // names
+  std::string name_;
+  std::string group_;
+
+  // tool link and ik tolerance
+  std::string tool_link_;
```

Comment:
This is not initialized in constructor / init function.

Reply:
I move initialization of tool_link_ to method configure and I define ros param